api/client/api_client.go

```diff
@@ -0,0 +1,139 @@
+	return fmt.Sprintf("http://%v:%v/%v/%v", client.ipAddr, client.port, personEndpoint, id)
+}
+
+func (client *APIClient) AddPerson(id int) error {
+	url := client.getPersonUrlForId(id)
+	resp, err := client.httpClient.Post(url, textContentType, nil)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred making the request to add person with ID '%v'", id)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return stacktrace.NewError("Adding person with ID '%v' returned non-OK status code %v", id, resp.StatusCode)
+	}
+	return nil
+}
+
+func (client *APIClient) GetPerson(id int) (Person, error) {
+	url := client.getPersonUrlForId(id)
+	resp, err := client.httpClient.Get(url)
+	if err != nil {
+		return Person{}, stacktrace.Propagate(err, "An error occurred making the request to get person with ID '%v'", id)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return Person{}, stacktrace.NewError("Getting person with ID '%v' returned non-OK status code %v", id, resp.StatusCode)
+	}
+	body := resp.Body
+	defer body.Close()
+	bodyBytes, err := ioutil.ReadAll(body)
+	if err != nil {
+		return Person{}, stacktrace.Propagate(err, "An error occurred reading the response body")
+	}
+
+	var person Person
+	if err := json.Unmarshal(bodyBytes, &person); err != nil {
+		return Person{}, stacktrace.Propagate(err, "An error occurred deserializing the Person JSON")
+	}
+	return person, nil
+}
+
+func (client *APIClient) IncrementBooksRead(id int) error {
+	url := fmt.Sprintf("http://%v:%v/%v/%v", client.ipAddr, client.port, incrementBooksReadEndpoint, id)
+	resp, err := client.httpClient.Post(url, textContentType, nil)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred making the request to increment the books read of person with ID '%v'", id)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return stacktrace.NewError("Incrementing the books read of person with ID '%v' returned non-OK status code %v", id, resp.StatusCode)
+	}
+	return nil
+}
+
+/*
+Wait for healthy response
+*/
+func (client *APIClient) WaitForHealthy(retries uint32, retriesDelayMilliseconds int) error {
+
+	var(
+		url = fmt.Sprintf("http://%v:%v/%v", client.ipAddr, client.port, healthcheckUrlSlug)
+		resp *http.Response
+		err error
+	)
+
+	for i := uint32(0); i < retries; i++ {
+		resp, err = client.makeHttpGetRequest(url)
+		if err == nil  {
+			break
+		}
+		time.Sleep(time.Duration(retriesDelayMilliseconds) * time.Millisecond)
+	}
+
+	body := resp.Body
+	defer body.Close()
+
+	bodyBytes, err := ioutil.ReadAll(body)
+	if err != nil {
+		return stacktrace.NewError("An error occurred reading the response body: %v", err)
+	}
+	bodyStr := string(bodyBytes)
+
+	if bodyStr != healthyValue {
+		return stacktrace.NewError("Expected response body text '%v' from endpoint '%v' but got '%v' instead", healthyValue, url, bodyStr)
+	}
+
+	return nil
+}
+
+func (client *APIClient) makeHttpGetRequest(url string) (*http.Response, error){
```

Comment:
Normally I wouldn't like copy-pasting the same code in two places, but in this case the API service & datastore service are technically separate, isolated things and don't "know" about each other so this makes sense here

Reply:
Yes, I agree, I usually try to follow the DRY principle, but on this occasion makes sense to have both.